File: export_player_template/runtime_persistent_variables.js
export const RUNTIME_VARIABLE_SCOPES = Object.freeze(["save", "persistent"]);
export const PERSISTENT_VARIABLE_STORE_FORMAT_VERSION = 1;

export function getSafeRuntimeVariableScope(value) {
  return String(value ?? "").trim().toLowerCase() === "persistent" ? "persistent" : "save";
}

export function isPersistentRuntimeVariable(variable) {
  return getSafeRuntimeVariableScope(variable?.scope) === "persistent";
}

export function getPersistentRuntimeVariables(variables = []) {
  return (Array.isArray(variables) ? variables : []).filter(
    (variable) => variable?.id && isPersistentRuntimeVariable(variable)
  );
}

function getSafeNumber(value, fallback = 0) {
  const parsed = typeof value === "number" ? value : Number.parseFloat(value ?? "");
  return Number.isFinite(parsed) ? parsed : fallback;
}

function getNumberBound(value) {
  if (value === null || value === undefined || typeof value === "boolean") {
    return null;
  }
  const parsed = Number.parseFloat(value);
  return Number.isFinite(parsed) ? parsed : null;
}

export function coercePersistentRuntimeVariableValue(variable, value) {
  const rawType = String(variable?.type ?? "string").trim().toLowerCase();
  const type = rawType === "number" || rawType === "boolean" ? rawType : "string";
  const fallback = variable?.defaultValue;

  if (type === "number") {
    let nextValue = getSafeNumber(value, getSafeNumber(fallback, 0));
    const minValue = getNumberBound(variable?.min ?? variable?.minValue);
    const maxValue = getNumberBound(variable?.max ?? variable?.maxValue);
    if (minValue !== null) {
      nextValue = Math.max(nextValue, minValue);
    }
    if (maxValue !== null) {
      nextValue = Math.min(nextValue, maxValue);
    }
    return nextValue;
  }

  if (type === "boolean") {
    if (typeof value === "boolean") {
      return value;
    }
    if (typeof value === "string") {
      const normalized = value.trim().toLowerCase();
      if (["true", "1", "yes", "on"].includes(normalized)) return true;
      if (["false", "0", "no", "off", ""].includes(normalized)) return false;
    }
    return typeof fallback === "boolean" ? fallback : Boolean(value);
  }

  return value === null || value === undefined
    ? typeof fallback === "string"
      ? fallback
      : ""
    : String(value);
}

function getPersistentValueSource(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }
  return value.values && typeof value.values === "object" && !Array.isArray(value.values)
    ? value.values
    : value;
}

export function sanitizePersistentRuntimeVariableState(value, variables = []) {
  const source = getPersistentValueSource(value);
  return getPersistentRuntimeVariables(variables).reduce((result, variable) => {
    const rawValue = Object.hasOwn(source, variable.id) ? source[variable.id] : variable.defaultValue;
    result[variable.id] = coercePersistentRuntimeVariableValue(variable, rawValue);
    return result;
  }, {});
}

export function mergePersistentRuntimeVariableState(variableState, variables = [], persistentState = {}) {
  return {
    ...(variableState && typeof variableState === "object" ? variableState : {}),
    ...sanitizePersistentRuntimeVariableState(persistentState, variables),
  };
}

export function collectPersistentRuntimeVariableState(variableState, variables = []) {
  return sanitizePersistentRuntimeVariableState(variableState, variables);
}

export function buildPersistentRuntimeVariableStore(variableState, variables = [], options = {}) {
  const now = typeof options.now === "function" ? options.now() : new Date().toISOString();
  return {
    formatVersion: PERSISTENT_VARIABLE_STORE_FORMAT_VERSION,
    updatedAt: String(now || ""),
    values: collectPersistentRuntimeVariableState(variableState, variables),
  };
}

export function getPersistentRuntimeVariableSummary(variableState, variables = []) {
  const definitions = getPersistentRuntimeVariables(variables);
  const values = sanitizePersistentRuntimeVariableState(variableState, variables);
  return {
    count: definitions.length,
    values,
    changedCount: definitions.filter(
      (variable) =>
        JSON.stringify(values[variable.id]) !==
        JSON.stringify(coercePersistentRuntimeVariableValue(variable, variable.defaultValue))
    ).length,
  };
}

File: native_runtime/runtime_persistent_variables.py
from __future__ import annotations

from typing import Any


RUNTIME_VARIABLE_SCOPES = ("save", "persistent")
PERSISTENT_VARIABLE_STORE_FORMAT_VERSION = 1


def get_safe_runtime_variable_scope(value: object) -> str:
    return "persistent" if str(value or "").strip().lower() == "persistent" else "save"


def is_persistent_runtime_variable(variable: dict | None) -> bool:
    return get_safe_runtime_variable_scope((variable or {}).get("scope")) == "persistent"


def get_persistent_runtime_variables(variables: list[dict] | None) -> list[dict]:
    return [
        variable
        for variable in variables or []
        if isinstance(variable, dict) and variable.get("id") and is_persistent_runtime_variable(variable)
    ]


def _safe_number(value: object, fallback: float = 0) -> float | int:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        parsed = float(fallback)
    return int(parsed) if parsed.is_integer() else parsed


def _number_bound(value: object) -> float | None:
    if value is None or isinstance(value, bool):
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def coerce_persistent_runtime_variable_value(variable: dict, value: object) -> object:
    variable_type = str(variable.get("type") or "string").strip().lower()
    fallback = variable.get("defaultValue")

    if variable_type == "number":
        next_value = _safe_number(value, _safe_number(fallback, 0))
        min_value = _number_bound(variable.get("min", variable.get("minValue")))
        max_value = _number_bound(variable.get("max", variable.get("maxValue")))
        if min_value is not None:
            next_value = max(float(next_value), min_value)
        if max_value is not None:
            next_value = min(float(next_value), max_value)
        return int(next_value) if float(next_value).is_integer() else next_value

    if variable_type == "boolean":
        if isinstance(value, bool):
            return value
        if isinstance(value, str):
            normalized = value.strip().lower()
            if normalized in {"true", "1", "yes", "on"}:
                return True
            if normalized in {"false", "0", "no", "off", ""}:
                return False
        return fallback if isinstance(fallback, bool) else bool(value)

    if value is None:
        return fallback if isinstance(fallback, str) else ""
    return str(value)


def _persistent_value_source(value: object) -> dict:
    if not isinstance(value, dict):
        return {}
    nested_values = value.get("values")
    return nested_values if isinstance(nested_values, dict) else value


def sanitize_persistent_runtime_variable_state(
    value: object,
    variables: list[dict] | None,
) -> dict[str, object]:
    source = _persistent_value_source(value)
    result: dict[str, object] = {}
    for variable in get_persistent_runtime_variables(variables):
        variable_id = str(variable.get("id") or "")
        raw_value = source.get(variable_id, variable.get("defaultValue"))
        result[variable_id] = coerce_persistent_runtime_variable_value(variable, raw_value)
    return result


def merge_persistent_runtime_variable_state(
    variable_state: dict | None,
    variables: list[dict] | None,
    persistent_state: object,
) -> dict[str, object]:
    merged = dict(variable_state) if isinstance(variable_state, dict) else {}
    merged.update(sanitize_persistent_runtime_variable_state(persistent_state, variables))
    return merged


def collect_persistent_runtime_variable_state(
    variable_state: dict | None,
    variables: list[dict] | None,
) -> dict[str, object]:
    return sanitize_persistent_runtime_variable_state(variable_state or {}, variables)


def build_persistent_runtime_variable_store(
    variable_state: dict | None,
    variables: list[dict] | None,
    *,
    updated_at: str = "",
) -> dict[str, Any]:
    return {
        "formatVersion": PERSISTENT_VARIABLE_STORE_FORMAT_VERSION,
        "updatedAt": str(updated_at or ""),
        "values": collect_persistent_runtime_variable_state(variable_state, variables),
    }


def get_persistent_runtime_variable_summary(
    variable_state: dict | None,
    variables: list[dict] | None,
) -> dict[str, Any]:
    definitions = get_persistent_runtime_variables(variables)
    values = sanitize_persistent_runtime_variable_state(variable_state or {}, variables)
    changed_count = sum(
        1
        for variable in definitions
        if values.get(str(variable.get("id") or ""))
        != coerce_persistent_runtime_variable_value(variable, variable.get("defaultValue"))
    )
    return {
        "count": len(definitions),
        "changedCount": changed_count,
        "values": values,
    }
